lib/nerve.rb

```diff
@@ -86,15 +165,47 @@ def run
       $EXIT = true
     end
 
-    def launch_watcher(name, config)
-      log.debug "nerve: launching service watcher #{name}"
-      watcher = ServiceWatcher.new(config.merge({'instance_id' => @instance_id, 'name' => name}))
-      @watchers[name] = Thread.new{watcher.run}
+    def merged_config(config, name)
+      # Get a deep copy so sub-hashes are properly handled
+      deep_copy = Marshal.load(Marshal.dump(config))
+      return deep_copy.merge({'instance_id' => @instance_id, 'name' => name})
+    end
+
+    def launch_watcher(name, config, wait=false)
+      watcher_config = merged_config(config, name)
+      # The ServiceWatcher may mutate the configs, so record the version before
+      # passing the config to the ServiceWatcher
+      @watcher_versions[name] = watcher_config.hash
+
+      watcher = ServiceWatcher.new(watcher_config)
+      unless @config_manager.options[:check_config]
+        log.debug "nerve: launching service watcher #{name}"
+        watcher_thread = Thread.new { watcher.run }
```

Comment:
maybe one way to simplify the testing is to move the thread management into the `ServiceWatcher` class? then, in the tests here, you simply assert that you call `watcher.run` or `watcher.stop`. you can test the thread-specific functionality using tests on the watcher -- including internal coordination.

you could then also avoid using thread-local variables, and instead use instance variables on the watcher instance. so, for instance, you would call `watcher.has_reported?` instead of checking `watcher_thread[:has_reported]`. internally, instead of `Thread.current[:has_reported]`, that method would just return `@has_reported`